src/render/locals.rs

```diff
@@ -89,87 +91,148 @@ pub fn render_locals<'a, 'tcx: 'a>(ecx: &EvalContext<'a, 'tcx>, frame: Option<&F
+            if let Value::Scalar(Scalar::Bits { defined: 0, .. }) = val {
+                Err(EvalErrorKind::AssumptionNotHeld)?;
+            }
+
+            let variant = if adt_def.variants.len() == 1 {
+                0
+            } else if let Value::ByRef(_ptr, _align) = val {
+                // FIXME: read the discriminant in a safe way
+                // let ptr = ptr.to_ptr()?;
+                // unsafe { &mut *(ecx as *const EvalContext as *mut EvalContext) }.read_discriminant_as_variant_index(Place::from_ptr(ptr, align), ty)?
+                Err(EvalErrorKind::AssumptionNotHeld)?
+            } else {
+                Err(EvalErrorKind::AssumptionNotHeld)?
+            };
+
+            //println!("{:?} {:?} {:?}", val, ty, adt_def.variants);
+            let mut pretty = format!("{} {{ ", ecx.tcx.absolute_item_path_str(adt_def.did).replace("<", "&lt;").replace(">", "&gt;"));
+            for (i, adt_field) in adt_def.variants[variant].fields.iter().enumerate() {
+                let field_pretty: EvalResult<String> = do catch {
+                    let ValTy { value: field_val, ty: field_ty } = ecx.read_field(val, None, ::rustc::mir::Field::new(i), ty)?;
+                    pp_value(ecx, field_ty, field_val)?
+                };
+                pretty.push_str(&format!("{}: {}, ", adt_field.ident.as_str(), match field_pretty {
+                    Ok(field_pretty) => field_pretty,
+                    Err(_err) => "<span style='color: red;'>&lt;err&gt;</span>".to_string(),
+                }));
+            }
+            pretty.push_str("}");
+            println!("pretty adt: {}", pretty);
+            return Ok(pretty);
+        }
+        _ => {}
+    }
+
+    if layout.size.bytes() == 0 {
+        Err(EvalErrorKind::AssumptionNotHeld)?;
+    }
+    if let Abi::Scalar(_) = layout.abi {
+    } else {
+        Err(EvalErrorKind::AssumptionNotHeld)?;
+    }
+    let scalar = ecx.value_to_scalar(ValTy { value: val, ty })?;
+    if let Scalar::Ptr(_) = &scalar {
+        return Ok(print_primval(scalar)); // If the value is a ptr, print it
+    }
+    let bits = scalar.to_bits(layout.size)?;
+    match ty.sty {
+        TypeVariants::TyBool => {
+            if bits == 0 {
+                Ok("false".to_string())
+            } else if bits == 1 {
+                Ok("true".to_string())
+            } else {
+                Err(EvalErrorKind::AssumptionNotHeld.into())
+            }
+        }
+        TypeVariants::TyChar if bits < ::std::char::MAX as u128 => {
+            let chr = ::std::char::from_u32(bits as u32).unwrap();
+            if chr.is_ascii() {
+                Ok(format!("'{}'", chr))
+            } else {
+                Err(EvalErrorKind::AssumptionNotHeld.into())
+            }
+        }
+        TypeVariants::TyUint(_) => {
+            Ok(format!("{0}", bits))
+        }
+        TypeVariants::TyInt(_) => {
+            Ok(format!("{0}", bits as i128))
```

Comment:
this should not work for `i64` and smaller. You need to sign extend first

Reply:
Fixed